File: 020.cc
#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
    void push(int value) {
        this->m_data.push_back(value);
    }
    void pop() {
        this->m_data.erase(this->m_data.end() - 1);
    }
    int top() {
        if (0 == this->m_data.size()) return int(NULL);
        return this->m_data.at(0);
    }
    int min() {
        if (0 == this->m_data.size()) return int(NULL);
        int m = this->top();
        for (int i=1; i<this->m_data.size(); i++)
            if (m > this->m_data.at(i))
                m = this->m_data.at(i);
        return m;
    }
private:
    vector<int> m_data;
};

int main(int argc, char const *argv[])
{
    Solution * p = new Solution();
    p->push(3);
    cout << p->min() << endl;
    p->push(4);
    cout << p->min() << endl;
    p->push(2);
    cout << p->min() << endl;
    p->push(3);
    cout << p->min() << endl;
    p->pop();
    cout << p->min() << endl;
    p->pop();
    cout << p->min() << endl;
    p->push(0);
    cout << p->min() << endl;
    return 0;
}

File: python/020.py
# -*- coding: utf-8 -*-
# /usr/bin/python3
# Creator: Hemajun
# Description: 定义栈的数据结构，请在该类型中实现一个能够得到栈最小元素的min函数。


class Solution:
    def __init__(self):
        self.__data = []
        self.__min = None

    def push(self, node):
        self.__data.append(node)
        self.__min = node if self.__min is None or self.__min > node else self.__min

    def pop(self):
        result = None
        if self.__data:
            result = self.__data.pop()
            self.__min = min(
                self.__data) if result == self.__min else self.__min
        return result

    def top(self):
        return self.__data[0] if self.__data else None

    def min(self):
        return self.__min
